### ISSUE TITLE
ERL-1476: xref erroneously claims M:behaviour_info/1 is undefined

### ISSUE THREAD

**Issue by OTP-Maintainer** [erlang/otp#4192]
Original reporter: `mikael pettersson`
Affected versions: `OTP-24.0`, `OTP-23.2`
Component: `Not Specified`
Migrated from: https://bugs.erlang.org/browse/ERL-1476

---

```
Upgrading from OTP-23.1.5 to OTP-23.2.3 caused our xref checks to fail, with complaints that calls to M:behaviour_info/1 are undefined, even though they provably are not. A git bisect identified the culprit:

{noformat}
# first bad commit: [d74e9b343a8562aa2ffceb07ad44a1fa8fd45f48] Suppress xref output for unused export on behaviour's behaviour_info/1
{noformat}

That's [https://github.com/erlang/otp/pull/2752|https://github.com/erlang/otp/pull/2752.] aka https://bugs.erlang.org/browse/ERL-1353, which apparently wanted to prevent xref from reporting compiler-generated behaviour_info/1 functions as unused exports. Fine, but the implementation appears too crude, resulting in the problem reported here.

Repeat-by:
{noformat}
> cat a.erl
-module(a).
-export([foo/1]).
-callback foo() -> ok.
foo(Mod) -> Mod:foo().
> cat b.erl
-module(b).
-export([bar/0]).
bar() -> a:behaviour_info(callbacks).
> cat x.erl
-module(x).
-export([ref/0]).
ref() ->
  {ok, XREF} = xref:start([]),
  ok = xref:set_default(XREF, [{warnings, false}]),
  {ok, _} = xref:add_module(XREF, "a.beam"),
  {ok, _} = xref:add_module(XREF, "b.beam"),
  {ok, UndefCalls} = xref:analyze(XREF, undefined_function_calls),
  case UndefCalls of
    [] -> io:format(standard_io, "No undefined function calls\n", []);
    _ -> io:format(standard_io, "Undefined function calls: ~p\n", [UndefCalls])
  end.
{noformat}
{{a}} implements a behaviour, {{b}} a test case, and {{x}} an {{xref}} wrapper. Let's run with OTP-23.2.3:
{noformat}
> erlc +debug_info a.erl b.erl x.erl
> erl -noshell -s x ref -s erlang halt
Undefined function calls: [{{b,bar,0},{a,behaviour_info,1}}]
{noformat}
This is is clearly wrong. The result is the same with current master. A re-run with OTP-23.1.5 shows the expected result:
{noformat}
> otp.git/bin/erlc +debug_info a.erl b.erl x.erl
> otp.git/bin/erl -noshell -s x ref -s erlang halt
No undefined function calls
{noformat}

I'll note that https://erlang.org/doc/design_principles/spec_proc.html#behaviours is pretty clear in implying that M:behaviour_info/1 is expected to exist.
```

**Comment by uabboli** [comment]
Thanks for pointing out the bug. I believe the linked PR solves the problem.

**Comment by paulo-ferraz-oliveira** [comment]
Oops 😞. I guess I didn't add enough tests. I'm sorry for that. As stated, the only goal was to suppress a warning, not remove functionality.

@uabboli: is the main error in my implementation the fact that I didn't use `erl_anno:location(Anno)`? (I can't remember if `{Line, Column}` was already available, at the time)

i.e.

```erlang
form({function, 0 = _Line, behaviour_info = _Name, 1 = _Arity, _Clauses}, S) ->
    S;
```

v.

```erlang
form({function, Anno, Name, Arity, Clauses}, S) ->
    case {Name, Arity, erl_anno:location(Anno)} of
        {behaviour_info, 1, 0} ->
            S; % generated
```

(added some _-prefixed variables for clarity, here)

**Comment by paulo-ferraz-oliveira** [comment]
I believe there's something wrong (as stated), but have yet not experienced it with a recent version from `master` and I've ran a lot of XRef analysis; I'm try again, since it's possible it's due to different options, for example.

**Comment by paulo-ferraz-oliveira** [comment]
The issue seems to be fixed in this commit, at least.

**Comment by uabboli** [comment]
Merged commit ba9f40fb710f3f7b908b045e62ffd07b4186dcc3 to erlang:maint-23 (Erlang OTP 23.2.5).
